### PR TITLE
tests: core20 preseed/nested spread test

### PR DESCRIPTION
**Pull request by stolowski**
Nested spread for uc20 preseeding.

Getting `NESTED_UBUNTU_IMAGE_PRESEED_KEY` through to nested.sh helper quoted correctly was tricky and shellcheck doesn't like this variable (complains it's unused and also complains about parenthesis) therefore this test is excluded from shellcheck (`#shellcheck disable..` didn't have any effect on environment block in my tests)

### CODE REVIEW

File: run-checks [hunk 1]
```diff
@@ -218,7 +218,7 @@ if [ "$STATIC" = 1 ]; then
         fi
         unset regexp
         # also run spread-shellcheck
-        ./tests/lib/external/snapd-testing-tools/utils/spread-shellcheck spread.yaml tests --exclude "$exclude_tools_path"
+        ./tests/lib/external/snapd-testing-tools/utils/spread-shellcheck spread.yaml tests --exclude "$exclude_tools_path" --exclude "tests/nested/manual/core20-preseed"
```
Comment (sergiocazzolato) [inline run-checks:221]: could you please add a comment explaining why it needs to be excluded?

File: tests/nested/manual/core20-preseed/task.yaml [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  echo "Expose the needed assertions through the fakestore"
+  cp "$TESTSLIB"/assertions/developer1.account "$NESTED_FAKESTORE_BLOB_DIR"/asserts
+  cp "$TESTSLIB"/assertions/developer1.account-key "$NESTED_FAKESTORE_BLOB_DIR"/asserts
+  cp "$TESTSLIB"/assertions/testrootorg-store.account-key "$NESTED_FAKESTORE_BLOB_DIR"/asserts
+
+  teardown_staging_store
+
+  echo "Creating a new key without a password"
+  mkdir -p ~/.snap/gnupg
+  gendeveloper1 show-key | gpg --homedir=~/.snap/gnupg --import
```
Comment (sergiocazzolato) [inline tests/nested/manual/core20-preseed/task.yaml:35]: We should remove this key in restore

**Review by sergiocazzolato — COMMENTED** [review]
Thanks for those tests, I left few comments inline

File: run-checks [hunk 3]
```diff
@@ -218,7 +218,8 @@ if [ "$STATIC" = 1 ]; then
         fi
         unset regexp
         # also run spread-shellcheck
-        ./tests/lib/external/snapd-testing-tools/utils/spread-shellcheck spread.yaml tests --exclude "$exclude_tools_path"
+        # XXX: exclude core20-preseed test as its environment black confuses shellcheck, and it's not possible to disable shellcheck there.
```
Comment (bboozzoo) [inline run-checks:221]:
```suggestion
        # XXX: exclude core20-preseed test as its environment block confuses shellcheck, and it's not possible to disable shellcheck there.
```

File: tests/lib/nested.sh [hunk 4]
```diff
@@ -797,13 +798,25 @@ EOF
...
+            else
+                # ubuntu-image creates sparse image files
+                # shellcheck disable=SC2086
+                "$UBUNTU_IMAGE" snap --image-size 10G "$NESTED_MODEL" \
+                    $UBUNTU_IMAGE_CHANNEL_ARG \
+                    --output-dir "$NESTED_IMAGES_DIR" \
+                    $EXTRA_FUNDAMENTAL \
+                    $EXTRA_SNAPS
+            fi
+
```
Comment (bboozzoo) [inline tests/lib/nested.sh:819]: would this work or will quoting break it?
```suggestion
            if [ -n "$NESTED_UBUNTU_IMAGE_PRESEED_KEY" ]; then
                UBUNTU_IMAGE_PRESEED_ARG=--preseed --preseed-sign-key=\""$NESTED_UBUNTU_IMAGE_PRESEED_KEY"\" 
            fi
            # ubuntu-image creates sparse image files
            # shellcheck disable=SC2086
            "$UBUNTU_IMAGE" snap --image-size 10G "$NESTED_MODEL" \
                $UBUNTU_IMAGE_CHANNEL_ARG \
                $UBUNTU_IMAGE_PRESEED_ARG \
                --output-dir "$NESTED_IMAGES_DIR" \
                $EXTRA_FUNDAMENTAL \
                $EXTRA_SNAPS
```

Comment (stolowski) [reply tests/lib/nested.sh:819]: Unfortunately it fails with:
```
/home/gopath/src/github.com/snapcore/snapd/tests/lib/nested.sh: line 803: --preseed-sign-key=" (test)": command not found
```

Trying to quote this assignment makes it fail when calling actual ubuntu-image.

Comment (mardy) [reply tests/lib/nested.sh:819]: This should work:
```bash
declare -a UBUNTU_IMAGE_PRESEED_ARGS
if [ -n "$NESTED_UBUNTU_IMAGE_PRESEED_KEY" ]; then
    UBUNTU_IMAGE_PRESEED_ARGS+=(--preseed  --preseed-sign-key=\""$NESTED_UBUNTU_IMAGE_PRESEED_KEY"\")
fi
"$UBUNTU_IMAGE" snap --image-size 10G "$NESTED_MODEL" \
                $UBUNTU_IMAGE_CHANNEL_ARG \
                "${UBUNTU_IMAGE_PRESEED_ARGS[@]}" \
                --output-dir "$NESTED_IMAGES_DIR" \
                $EXTRA_FUNDAMENTAL \
                $EXTRA_SNAPS
```

Comment (stolowski) [reply tests/lib/nested.sh:819]: Interesting, will try, thanks!

Comment (stolowski) [reply tests/lib/nested.sh]: Yeah, it worked! Thanks!

File: tests/lib/nested.sh [hunk 5]
```diff
@@ -797,13 +798,20 @@ EOF
             else 
                 UBUNTU_IMAGE_CHANNEL_ARG=""
             fi
+
+            declare -a UBUNTU_IMAGE_PRESEED_ARGS
```
Comment (sergiocazzolato) [inline tests/lib/nested.sh:802]: you could use here UBUNTU_IMAGE_PRESEED_ARGS as a string
UBUNTU_IMAGE_PRESEED_ARGS="" and then in the if
UBUNTU_IMAGE_PRESEED_ARGS="--preseed  --preseed-sign-key=\"$NESTED_UBUNTU_IMAGE_PRESEED_KEY\""
So this is more similar as we use the other variables.

Comment (stolowski) [reply tests/lib/nested.sh:802]: That didn't work; and I tried something similar before. With the change you suggested I'm getting the following (note the key value is not quoted and has a space in front):

+ UBUNTU_IMAGE_PRESEED_ARGS='--preseed --preseed-sign-key= (test)'
+ /home/gopath/bin/ubuntu-image snap --image-size 10G /home/gopath/src/github.com/snapcore/snapd/tests/lib/assertions/developer1-20-dangerous.model --channel edge --preseed --preseed-sign-key= '(test)' --output-dir /tmp/work-dir/images --snap /tmp/work-dir/assets/pc-kernel_5.4.0-113.127.1_amd64.snap --snap /tmp/work-dir/assets/pc_20-0.4_amd64.snap --snap /home/gopath/src/github.com/snapcore/snapd/tests/nested/manual/core20-preseed/snapd-from-deb.snap --snap /home/gopath/src/github.com/snapcore/snapd/tests/nested/manual/core20-preseed/new-core20.snap

Quoting the entire $UBUNTU_IMAGE_PRESEED_ARGS when invoking u-i breaks args.

**Review by sergiocazzolato — APPROVED** [review]
Thanks

**Review by bboozzoo — APPROVED** [review]
LGTM

File: tests/nested/manual/core20-preseed/task.yaml [hunk 6]
```diff
@@ -0,0 +1,69 @@
+summary: Check that preseeding of core20 works and the image boots.
+
+details: |
+  This test checks that preseeding of UC20 image with ubuntu-image works
+  and that the resulting image boots.
+
+systems: [ubuntu-20.04-64]
```
Comment (mvo5) [inline tests/nested/manual/core20-preseed/task.yaml:7]: Do we need something similar for uc22 (in a followup?)

**Review by mvo5 — APPROVED** [review]

**Pull request merged**